### PR TITLE
feat(sgx): SecInfoExt

### PR DESCRIPTION
**Pull request by jarkkojs**
```
Add SecInfoExt trait that provides EACCEPT, EACCEPTCOPY and EMODPE
operations for enclave's use.

Closes: #1519
Signed-off-by: Jarkko Sakkinen <jarkko@profian.com>
```

### CODE REVIEW

**Comment by jarkkojs** [comment]
Would it make sense to add `unsafe fn enclu()` to the trait to avoid repetition?

File: internal/shim-sgx/src/handler/mod.rs [hunk 1]
```diff
@@ -71,6 +73,63 @@ impl<'a> Write for Handler<'a> {
     }
 }
 
+trait SecInfoExt {
+    fn accept(&self, dest: *const u8);
+    fn accept_copy(&self, dest: *const u8, src: *const u8);
+    fn protect(&self, dest: *const u8);
```
Comment (haraldh) [inline internal/shim-sgx/src/handler/mod.rs:79]: do these accept any pointer or only page aligned ones?

Comment (haraldh) [reply internal/shim-sgx/src/handler/mod.rs:79]: Shouldn't they return the error code from `rax` ??
Best to create consts for them, too.

Comment (jarkkojs) [reply internal/shim-sgx/src/handler/mod.rs:79]: Oops, this is not the latest version. In the newest I uses NonNull. I'll update.

File: internal/shim-sgx/src/handler/mod.rs [hunk 2]
```diff
@@ -71,6 +73,63 @@ impl<'a> Write for Handler<'a> {
     }
 }
 
+trait SecInfoExt {
+    fn accept(&self, dest: NonNull<libc::c_void>);
+    fn accept_copy(&self, dest: NonNull<libc::c_void>, src: NonNull<libc::c_void>);
+    fn protect(&self, dest: NonNull<libc::c_void>);
```
Comment (haraldh) [inline internal/shim-sgx/src/handler/mod.rs:79]: Shouldn't they return the error code from rax ??

**Review by haraldh — CHANGES_REQUESTED** [review]

File: internal/shim-sgx/src/handler/mod.rs (hunk 2 above)
Comment (jarkkojs) [reply internal/shim-sgx/src/handler/mod.rs:79]: * EACCEPT, EACCEPTCOPY: yes
* EMODPE: no

Comment (jarkkojs) [reply internal/shim-sgx/src/handler/mod.rs]: I fixed this issue and pushed a new commit.

**Comment by jarkkojs** [comment]
Could this be merged soon, or is there still something to fix?

**Comment by haraldh** [comment]
what about pointer page alignment? check or not?

**Comment by jarkkojs** [comment]
> what about pointer page alignment? check or not?

I don't know where I messed things up but now the *const u8's are back to NonNull (just did a new push).

So, there could be page alignment check, yes, but does it matter? CPU will assert exception and thread will thrown out if that ever happens.

**Comment by jarkkojs** [comment]
I updated the commit:
- Use NonNull.
- Return usize instead of Result<usize>. SGX error code is a legit execution i.e. there are only legit results from these kind of.

**Comment by jarkkojs** [comment]
New version pushed now the check is:
```rust
    if std::path::Path::new("/dev/sgx_enclave").exists()
        && fs::metadata("/dev/sgx_enclave")
            .unwrap()
            .file_type()
            .is_char_device()
    {
        println!("cargo:rustc-cfg=host_can_test_sgx");

        if std::path::Path::new(AESM_SOCKET).exists()
            && fs::metadata(AESM_SOCKET).unwrap().file_type().is_socket()
        {
            println!("cargo:rustc-cfg=host_can_test_attestation");
        }
    }
```

**Pull request closed without merging**